garak/generators/llm.py

```diff
@@ -0,0 +1,120 @@
+
+    Explicitly, garak delegates the majority of responsibility here:
+
+    * the generator calls prompt() on the resolved `llm` model
+    * provider setup, auth, and model-specific options live in `llm`
+    * there's no support for chains; this is a direct LLM interface
+
+    Notes:
+    * Not all providers support all parameters (e.g., temperature, max_tokens).
+      We pass only non-None params; providers ignore what they don't support.
+    """
+
+    DEFAULT_PARAMS = Generator.DEFAULT_PARAMS | {
+        "max_tokens": None,
+        "top_p": None,
+        "stop": [],
+    }
+
+    generator_family_name = "llm"
+
+    def __init__(self, name: str = "", config_root=_config):
+        self.target = None
+        self.name = name
+        self._load_config(config_root)
+        self.fullname = f"llm (simonw/llm) {self.name or '(default)'}"
+
+        self._load_client()
+
+        super().__init__(self.name, config_root=config_root)
+
+        self._clear_client()
+
+    def __getstate__(self) -> object:
+        self._clear_client()
+        return dict(self.__dict__)
+
+    def __setstate__(self, data: dict) -> None:
+        self.__dict__.update(data)
+        self._load_client()
+
+    def _load_client(self) -> None:
+        try:
+            self.target = llm.get_model(self.name) if self.name else llm.get_model()
+        except Exception as exc:
+            logging.error(
+                "Failed to resolve `llm` target '%s': %s", self.name, repr(exc)
+            )
+            raise
+
+    def _clear_client(self) -> None:
+        self.target = None
+
+    def _call_model(
+        self, prompt: Conversation, generations_this_call: int = 1
+    ) -> List[Union[Message, None]]:
+        """
+        Continuation generation method for LLM integrations via `llm`.
+
+        This calls model.prompt() once per generation and materializes the text().
+        """
+        if self.target is None:
+            self._load_client()
+
+        system_turns = [turn for turn in prompt.turns if turn.role == "system"]
+        user_turns = [turn for turn in prompt.turns if turn.role == "user"]
+        assistant_turns = [turn for turn in prompt.turns if turn.role == "assistant"]
+
+        if assistant_turns:
+            raise ValueError("llm generator does not accept assistant turns")
+        if len(system_turns) > 1:
+            raise ValueError("llm generator supports at most one system turn")
+        if len(user_turns) != 1:
+            raise ValueError("llm generator requires exactly one user turn")
+
+        text_prompt = prompt.last_message("user").text
+
+        # Build kwargs only for parameters explicitly set (non-None / non-empty)
+        prompt_kwargs = {
+            key: getattr(self, key)
+            for key in ("max_tokens", "temperature", "top_p")
+            if getattr(self, key) is not None
+        }
+        if self.stop:
+            prompt_kwargs["stop"] = self.stop
```

Comment:
Could this inspect the accepted arguments to `self.target.prompt()` vs a hard coded list here?  Something similar exists in the OpenAICompatible class, where we collect all options set on the generator that the target API accepts.